packages/stryker/src/ScoreResultCalculator.ts

```diff
@@ -0,0 +1,125 @@
+import { MutantResult, MutantStatus, ScoreResult } from 'stryker-api/report';
+import { freezeRecursively } from './utils/objectUtils';
+import * as _ from 'lodash';
+import * as path from 'path';
+
+export default class ScoreResultCalculator {
+
+  static calculate(results: MutantResult[], basePath: string = ''): ScoreResult {
+    const name = this.determineCommonBasePath(results, basePath);
+    const childResults = this.calculateChildScores(results, name, basePath);
+    const numbers = this.calculateNumbers(results, childResults);
+    return freezeRecursively(_.assign(numbers, { name, childResults }));
+  }
+
+  private static compareScoreResults(a: ScoreResult, b: ScoreResult) {
+    const sortValue = (scoreResult: ScoreResult) => {
+      // Directories first
+      if (scoreResult.childResults.length) {
+        return `0${scoreResult.name}`;
+      } else {
+        return `1${scoreResult.name}`;
+      }
+    };
+    return sortValue(a).localeCompare(sortValue(b));
+  }
+
+  private static calculateChildScores(results: MutantResult[], name: string, basePath: string) {
+    const childrenBasePath = path.join(basePath, name) + path.sep;
+    const resultsByFiles = _.groupBy(results, result => result.sourceFilePath.substr(childrenBasePath.length));
+    const uniqueFiles = Object.keys(resultsByFiles);
+    if (uniqueFiles.length > 1) {
+      const filesGroupedByDirectory = _.groupBy(uniqueFiles, file => file.split(path.sep)[0]);
+      return Object.keys(filesGroupedByDirectory)
+        .map(directory => this.calculate(_.flatMap(filesGroupedByDirectory[directory], file => resultsByFiles[file]), childrenBasePath))
```

Comment:
Let's say we have mutant results: 

```json
[{
   "sourceFilePath": "a/b",
   ...
},{
   "sourceFilePath": "a/c",
  ...
},{
   "sourceFilePath": "b/c",
  ...
}]
```

Then `filesGroupedByDirectory` will look like:

```json
{ 
  "a": ["a/b" "a/c"],
  "b": ["b/c"]
}
```

Now if we for each file lookup the mutant results here:  `file => resultsGroupedByFiles[file]`, you get an array of arrays: `MutantResult[][]`. We want to flatMap that to a `MutantResult[]`